docs/design/include/mcmini/MCSharedLibraryWrappers.c

```diff
@@ -0,0 +1,394 @@
+  pthread_mutex_init_ptr     = &pthread_mutex_init;
+  pthread_mutex_lock_ptr     = &pthread_mutex_lock;
+  pthread_mutex_unlock_ptr   = &pthread_mutex_unlock;
+  sem_post_ptr               = &sem_post;
+  sem_wait_ptr               = &sem_wait;
+  sem_init_ptr               = &sem_init;
+  exit_ptr                   = &exit;
+  abort_ptr                  = &abort;
+//   pthread_barrier_init_ptr   = &pthread_barrier_init;
+//   pthread_barrier_wait_ptr   = &pthread_barrier_wait;
+//   pthread_rwlock_init_ptr    = &pthread_rwlock_init;
+//   pthread_rwlock_unlock_ptr  = &pthread_rwlock_unlock;
+//   pthread_rwlock_rdlock_ptr  = &pthread_rwlock_rdlock;
+//   pthread_rwlock_wrlock_ptr  = &pthread_rwlock_wrlock;
+  pthread_cond_init_ptr      = &pthread_cond_init;
+  pthread_cond_wait_ptr      = &pthread_cond_wait;
+  pthread_cond_signal_ptr    = &pthread_cond_signal;
+  pthread_cond_broadcast_ptr = &pthread_cond_broadcast;
+  sleep_ptr                  = &sleep;
+#endif
+}
+
+#if MC_SHARED_LIBRARY
+int
+pthread_create(pthread_t *pthread, const pthread_attr_t *attr,
+               void *(*routine)(void *), void *arg)
+{
+  if(mcmini_enabled) return mc_pthread_create(pthread, attr, routine, arg);
+  else {
+	  /*pthread_create_ptr = dlsym(RTLD_NEXT, "pthread_create");
+	  return pthread_create_ptr;*/
+	  int return_value =  __real_pthread_create(pthread, attr, routine, arg);
+	  return return_value;
+	}
+}
+
+int
+pthread_join(pthread_t pthread, void **result)
+{
+  if(mcmini_enabled) return mc_pthread_join(pthread, result);
+  else return __real_pthread_join(pthread, result);
+}
+
+int
+pthread_mutex_init(pthread_mutex_t *mutex,
+                   const pthread_mutexattr_t *mutexattr)
+{  if(mcmini_enabled) return mc_pthread_mutex_init(mutex, mutexattr);
+  else {
+	  //pthread_mutex_init_ptr   = dlsym(RTLD_NEXT, "pthread_mutex_init");
+	  //return pthread_mutex_init_ptr;
+	  int result =  __real_pthread_mutex_init(mutex, mutexattr);
+    return result;
+  }
+}
+
+int pthread_mutex_lock(pthread_mutex_t *mutex)
+{
+  if(mcmini_enabled) return mc_pthread_mutex_lock(mutex);
+  else {  
+	  int return_value = __real_pthread_mutex_lock(mutex);
+	  count_calls++;
+	  /*
+	   * Map the shared memory into the memory with MAP_FIXED_NOREPLACE
+	   * Recording phase.
+	   */
+	  long pagesize = sysconf(_SC_PAGESIZE);
+	  void *desired_address = (void *)((uintptr_t)base_address + 64*count_calls * sizeof(pthread_mutex_t));
+	  //printf("\n -- D.A after count calls %p ---", desired_address);
+	  desired_address = (void *)(((uintptr_t)desired_address / pagesize) * pagesize);
+	  //printf("\n -- D.A after page adjustment %p ---", desired_address);
+	  int retry_count = 0;
+	  size_t offset = sizeof(pthread_mutex_t);
+    while (1) {
+        // Map the shared memory into memory at the desired address
+        mutex = (pthread_mutex_t *)mmap(desired_address, sizeof(pthread_mutex_t), PROT_READ | PROT_WRITE, MAP_SHARED | MAP_FIXED_NOREPLACE, fd, 0);
```

Comment:
I suppose I'm a bit confused for two reasons. Firstly, the mutex doesn't have to be located in shared memory. The `mcmini` process doesn't care all that much where it gets located or what it is, so long as later `libmcmini.so` tells `mcmini` that the mutex exists and is named `pthread_mutex_t` 1, 2, etc.

Secondly, I'm confused why we're even attempting `MAP_FIXED_NOREPLACE` when all that appears to be happening here is that we're searching for some free region. Since your `Rec_list` holds onto these addresses anyway, it doesn't really matter where they go. Looking at [`mmap(2)`](https://man7.org/linux/man-pages/man2/mmap.2.html) gives us:

> If addr is NULL, then the kernel chooses the (page-aligned) address at which to create the mapping; this is the most portable method of creating a new mapping.

For debugging purposes, it is probably useful to have the addresses close by. For that, again the man page:

> If addr is not NULL, then the kernel takes it as a hint about where to place the mapping; on Linux, the kernel will pick a nearby page boundary (but always above or equal to the value specified by /proc/sys/vm/mmap_min_addr) and attempt to create the mapping there.  If another mapping already exists there, the kernel picks a new address that may or may not depend on the hint.  The address of the new mapping is returned as the result of the call.

I think the reliability with `mmap(2)` and  a `NULL` addr is better here than re-attempting an arbitrary amount of times with `MAP_FIXED_NO_REPLACE`.

Reply:
If we use MAP_FIXED_NOREPLACE at the time of restart we know at what address we have stored the list that have recording. But if we use mmap(2) at the the time of restart how do we know the starting address of the list? @gc00 @maxwellpirtle

Reply:
My understanding of DMTCP restarting means that "all of memory is restored" (plus maybe some caveats), which means that if the list start is some global variable simply re-accessing the variable gives us the address. Consequently, I think `malloc(3)1 would be favored no need for `mmap(2)`. Essentially I'm advocating that we can effectively treat `libmcmini.so as part of the application code, which will likewise access variables as normal etc. Thoughts @gc00?